### PR TITLE
adding initial iOS export script

### PR DESCRIPTION
**Pull request by tkothadev**
## What does this PR do?

This export script provides rudimentary support for iOS devices on the experimental macOS build of O3DE. The main functionality runs CMake commands to prepare assets and generate the XCode project file for the given project in the directory. This does not do the full export process however.

Due to the need to setup an AppleID, and a provisioning profile inside of XCode, this is currently handled manually in the Xcode IDE. Future work on this script would include utilizing `xcodebuild` to automate that portion of the process. For now, to get something in before release, we stop at just XCode project generation.

## How was this PR tested?

The script was tested on a local macOS machine with a verified iOS device. O3DE Atom Sampler project was used to verify the build process ran correctly. Outcome was a single xcode project file, from which the iOS app could be built and deployed to the device.

### CODE REVIEW

**Comment by tkothadev** [comment]
Note: I do not make use of CMake presets in this script, because I have found it to not be portable on a fresh install of mac laptops. It usually required tinkering with the engine setup to get the preset functionality working correctly, and I'd like to increase the probability of a successful first use of the script. As such, the commands remain as they are currently.

File: scripts/o3de/o3de/export_project.py [hunk 1]
```diff
@@ -38,7 +38,7 @@
     # Test if Ninja is available from the command line to determine the generator and multi-config capability
```
Comment (spham-amzn) [inline scripts/o3de/o3de/export_project.py:38]: I know that this is not part of the PR, but this is a copy/paste error and the description does not match the action.
```suggestion
    # Test if Xcode is available from the command line to determine the generator
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+def export_ios_xcode_project(ctx: exp.O3DEScriptExportContext,
+                             tools_build_folder: pathlib.Path,
+                             ios_build_folder: pathlib.Path,
+                             should_build_tools:bool = True,
+                             should_process_assets: bool = True,
+                             logger: logging.Logger|None = None):
+    """
+    This function serves as an initial exporter for project to the iOS platform. The steps in this code will generate
+    an XCode project file containing all necessary build information to produce iOS executables.
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:25]: FYI: `Xcode` is the proper casing, not `XCode` (https://developer.apple.com/xcode/) . Can you change it for the entire file?

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        logger.setLevel(logging.ERROR)
+
+    tools_build_folder_str = str(tools_build_folder)
+    ios_build_folder_str = str(ios_build_folder)
+
+    # Optionally build the toolchain needed to process the assets
+    if should_build_tools:
+        process_command(["cmake", "-B", tools_build_folder_str, '-G', "Xcode","-DLY_UNITY_BUILD=ON"], cwd=ctx.project_path)
+
+        process_command(f"cmake --build {tools_build_folder_str} --target Editor AssetProcessorBatch --config profile".split(),
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:52]: Why are we splitting a hard-coded string when we can just add the components individually like the above line?
```suggestion
        process_command("cmake", "--build", tools_build_folder_str, "--target", "Editor", "AssetProcessorBatch", "--config", "profile",
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                    cwd=ctx.project_path)
+        
+    # Optionally process the assets
+    if should_process_assets:
+        asset_processor_batch_path = tools_build_folder / 'bin/profile/AssetProcessorBatch'
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:57]: Use `exp.get_asset_processor_batch_path`
```suggestion
        asset_processor_batch_path = exp.get_asset_processor_batch_path(tools_build_folder, True)
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        process_command([ str(asset_processor_batch_path), '--platforms=ios' ,
+                        '--project-path', ctx.project_path ], cwd=ctx.project_path)
+
+    #generate the xcode project file for the O3DE project
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:61]:
```suggestion
    # Generate the xcode project file for the O3DE project
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        )
+
+        default_tools_path = o3de_context.project_path / 'build/Mac'
+        default_ios_path = o3de_context.project_path / 'build/iOS'
+        parser.add_argument(exp.CUSTOM_SCRIPT_HELP_ARGUMENT,default=False,action='store_true',help='Show this help message and exit.')
+        parser.add_argument('-bt', '--build-tools', default=True, action='store_true',
+                            help="Specifies whether to build O3DE toolchain executables. This will build AssetBundlerBatch, AssetProcessorBatch.")
+        parser.add_argument('-tbp', '--tools-build-path', type=pathlib.Path, default=default_tools_path,
+                                help='Designates where the build files for the O3DE toolchain are generated. If not specified, default is <o3de_project_path>/build/tools.')
+        parser.add_argument('-ibp', '--ios-build-path', type=pathlib.Path, default=default_tools_path,
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:94]:
```suggestion
        parser.add_argument('-ibp', '--ios-build-path', type=pathlib.Path, default=default_ios_path,
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                                help='Designates where the build files for the O3DE toolchain are generated. If not specified, default is <o3de_project_path>/build/tools.')
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:95]:
```suggestion
                                help=f'Designates where the build files for the game launcher are generated. If not specified, default is {default_ios_path}')
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                    add_help=False
+        )
+
+        default_tools_path = o3de_context.project_path / 'build/Mac'
+        default_ios_path = o3de_context.project_path / 'build/iOS'
+        parser.add_argument(exp.CUSTOM_SCRIPT_HELP_ARGUMENT,default=False,action='store_true',help='Show this help message and exit.')
+        parser.add_argument('-bt', '--build-tools', default=True, action='store_true',
+                            help="Specifies whether to build O3DE toolchain executables. This will build AssetBundlerBatch, AssetProcessorBatch.")
+        parser.add_argument('-tbp', '--tools-build-path', type=pathlib.Path, default=default_tools_path,
+                                help='Designates where the build files for the O3DE toolchain are generated. If not specified, default is <o3de_project_path>/build/tools.')
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:93]:
```suggestion
                                help=f'Designates where the build files for the O3DE toolchain are generated. If not specified, default is {default_tools_path}.')
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        parser.add_argument('-ibp', '--ios-build-path', type=pathlib.Path, default=default_tools_path,
+                                help='Designates where the build files for the O3DE toolchain are generated. If not specified, default is <o3de_project_path>/build/tools.')
+        parser.add_argument('-assets', '--should-process-assets', default=True, action='store_true',
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:96]: Having `store_true` and `default=True` makes this argument useless. There is no way to set to false.

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 10]
```diff
@@ -14,24 +14,24 @@
 from o3de.export_project import process_command
 import pathlib
 
-def export_ios_xcode_project(ctx: exp.O3DEScriptExportContext,
+def export_ios_Xcode_project(ctx: exp.O3DEScriptExportContext,
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:17]: The X should remain lowercase to be consistent with [PEP 8](https://pep8.org/#descriptive-naming-styles) which is the convention our python code uses
```suggestion
def export_ios_xcode_project(ctx: exp.O3DEScriptExportContext,
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    if not logger:
+        logger = logging.getLogger()
+        logger.setLevel(logging.ERROR)
+
+    tools_build_folder_str = str(tools_build_folder)
+    ios_build_folder_str = str(ios_build_folder)
+
+    # Optionally build the toolchain needed to process the assets
+    if should_build_tools:
+        process_command(["cmake", "-B", tools_build_folder_str, '-G', "Xcode","-DLY_UNITY_BUILD=ON"], cwd=ctx.project_path)
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:50]: The `LY_UNITY_BUILD=ON` argument is not needed.
It is the default value, it is only needed to turn it off.
```suggestion
        process_command(["cmake", "-B", tools_build_folder_str, '-G', "Xcode"], cwd=ctx.project_path)
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+        process_command(["cmake", "--build", tools_build_folder_str, "--target", "Editor", "AssetProcessorBatch", "--config", "profile"],
+                    cwd=ctx.project_path)
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:53]: The Editor doesn't need to be build to process Assets for iOS.
That should be removed as it would save build time for the user when exporting

```suggestion
        process_command(["cmake", "--build", tools_build_folder_str, "--target", "AssetProcessorBatch", "--config", "profile"],
                    cwd=ctx.project_path)
```

Also doesn't the `AssetBundlerBatch` need to be used here?

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    # Optionally process the assets
+    if not skip_asset_processing:
+        asset_processor_batch_path = exp.get_asset_processor_batch_path(tools_build_folder, True)
+        process_command([ str(asset_processor_batch_path), '--platforms=ios' ,
+                        '--project-path', ctx.project_path ], cwd=ctx.project_path)
+
+    # Generate the Xcode project file for the O3DE project
+    cmake_toolchain_path = ctx.engine_path / 'cmake/Platform/iOS/Toolchain_ios.cmake'
+
+    process_command(['cmake', '-B', ios_build_folder_str, '-G', "Xcode", f'-DCMAKE_TOOLCHAIN_FILE={str(cmake_toolchain_path)}', '-DLY_UNITY_BUILD=ON', '-DLY_MONOLITHIC_GAME=1'],
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:64]: repeat: `LY_UNITY_BUILD=ON` isn't needed since it is the default value
```suggestion
    process_command(['cmake', '-B', ios_build_folder_str, '-G', "Xcode", f'-DCMAKE_TOOLCHAIN_FILE={str(cmake_toolchain_path)}', '-DLY_MONOLITHIC_GAME=1'],
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                    description="Exports a project as a generated iOS Xcode project in the project directory. "
+                                "In order to use this script, the engine and project must be setup and registered beforehand. ",
+                    epilog=exp.CUSTOM_CMAKE_ARG_HELP_EPILOGUE,
+                    formatter_class=argparse.RawTextHelpFormatter,
+                    add_help=False
+        )
+
+        default_tools_path = o3de_context.project_path / 'build/Mac'
+        default_ios_path = o3de_context.project_path / 'build/iOS'
+        parser.add_argument(exp.CUSTOM_SCRIPT_HELP_ARGUMENT,default=False,action='store_true',help='Show this help message and exit.')
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:89]: extreme nitpick: missing spaces after commas
```suggestion
        parser.add_argument(exp.CUSTOM_SCRIPT_HELP_ARGUMENT, default=False, action='store_true', help='Show this help message and exit.')
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+                    prog=f'o3de.py export-project -es {__file__}',
+                    description="Exports a project as a generated iOS Xcode project in the project directory. "
+                                "In order to use this script, the engine and project must be setup and registered beforehand. ",
+                    epilog=exp.CUSTOM_CMAKE_ARG_HELP_EPILOGUE,
+                    formatter_class=argparse.RawTextHelpFormatter,
+                    add_help=False
+        )
+
+        default_tools_path = o3de_context.project_path / 'build/Mac'
+        default_ios_path = o3de_context.project_path / 'build/iOS'
```
Comment (AMZN-alexpete) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:88]: I think these build paths should match the default paths in the CMake presets which are:
```suggestion
        default_tools_path = o3de_context.project_path / 'build/mac_xcode'
        default_ios_path = o3de_context.project_path / 'build/ios_mono'
```
From: https://github.com/o3de/o3de/blob/development/cmake/Platform/Mac/CMakePresets.json#L76
And:
https://github.com/o3de/o3de/blob/development/cmake/Platform/iOS/CMakePresets.json#L28

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+#
+
+import argparse
+import logging
+import sys
+
+import o3de.export_project as exp
+from o3de.export_project import process_command
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:14]: One of these import statement aren't needed.

Either names from the module are being imported using the `from o3de.export_project import ...` syntax or the module itself is being import using the `import o3de.export_project ...` syntax, not both.

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    This function serves as an initial exporter for project to the iOS platform. The steps in this code will generate
+    an Xcode project file containing all necessary build information to produce iOS executables. 
+    In order to build and deploy projects, it is recommended to open the project file in Xcode to configure and build accordingly.
+
+    Note: In order to use this functionality, you must be running this script from a MacOS machine with a valid copy of Xcode
+
+    Instructions to handle iOS projects in Xcode will be provided soon.
+
+    :param ctx:                                     The O3DE Script context provided by the export-command
+    :param tools_build_folder:                      Optional build path to build the tools. (Will default to build/tools if not supplied)
```
Comment (AMZN-alexpete) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:33]:
```suggestion
    :param tools_build_folder:                      Optional build path to build the tools. (see default_tools_path below)
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py (hunk 10 above)
Comment (tkothadev) [reply scripts/o3de/ExportScripts/export_source_ios_xcode.py:17]: I must have been sleep deprived to let that sneak in... fixing it now

**Review by lumberyard-employee-dm — APPROVED** [review]

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 18]
```diff
@@ -47,21 +46,21 @@ def export_ios_Xcode_project(ctx: exp.O3DEScriptExportContext,
...
 
-        process_command(["cmake", "--build", tools_build_folder_str, "--target", "Editor", "AssetProcessorBatch", "--config", "profile"],
+        exp.process_command(["cmake", "--build", tools_build_folder_str, "--target", "AssetProcessorBatch", "AssetBundlerBatch", "--config", "profile"],
                     cwd=ctx.project_path)
         
     # Optionally process the assets
     if not skip_asset_processing:
         asset_processor_batch_path = exp.get_asset_processor_batch_path(tools_build_folder, True)
-        process_command([ str(asset_processor_batch_path), '--platforms=ios' ,
+        exp.process_command([ str(asset_processor_batch_path), '--platforms=ios' ,
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:57]:
```suggestion
        exp.process_command([ str(asset_processor_batch_path), '--platforms=ios',
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 19]
```diff
@@ -84,9 +83,9 @@ def parse_args(o3de_context: exp.O3DEScriptExportContext):
                     add_help=False
         )
 
-        default_tools_path = o3de_context.project_path / 'build/Mac'
-        default_ios_path = o3de_context.project_path / 'build/iOS'
-        parser.add_argument(exp.CUSTOM_SCRIPT_HELP_ARGUMENT,default=False,action='store_true',help='Show this help message and exit.')
+        default_tools_path = o3de_context.project_path / 'build/mac_xcode'
+        default_ios_path = o3de_context.project_path / 'build/ios_mono'
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:87]: The nomenclature here seems inconsistent. The `mac` uses a generator-specific suffix, while the `ios` uses a built type suffix. Why not use `tools` instead of `mac_xcode` for the tools path, and `game_ios` for the ios path?

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 20]
```diff
@@ -111,7 +110,7 @@ def parse_args(o3de_context: exp.O3DEScriptExportContext):
     if args.quiet:
         o3de_logger.setLevel(logging.ERROR)
     try:
-        export_ios_Xcode_project(ctx = o3de_context,
+        export_ios_xcode_project(ctx = o3de_context,
                             tools_build_folder= args.tools_build_path,
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:114]: nit: Can you align the additional args as well?

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py (hunk 19 above)
Comment (lumberyard-employee-dm) [reply scripts/o3de/ExportScripts/export_source_ios_xcode.py:87]: Well I guess it could be [ios_xcode_mono ](https://github.com/o3de/o3de/blob/development/cmake/Platform/iOS/CMakePresets.json#L28)

**Review by AMZN-alexpete — APPROVED** [review]

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py (hunk 19 above)
Comment (spham-amzn) [reply scripts/o3de/ExportScripts/export_source_ios_xcode.py:87]: i think `_mono` implies that you can have a non-mono, which isn't the case

Comment (lumberyard-employee-dm) [reply scripts/o3de/ExportScripts/export_source_ios_xcode.py:87]: On iOS it is possible to have non-mono.
It won't work, but the permutation is there.

Comment (spham-amzn) [reply scripts/o3de/ExportScripts/export_source_ios_xcode.py]: We actually do support non-monolithic for iOS. However, last time i checked, I think release builds disable any module loading anyways and will only work in monolithic mode. The other reason is for consistency, since both windows and linux do not use this nomenclature.

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    
+    args = parse_args(o3de_context)
+    if args.quiet:
+        o3de_logger.setLevel(logging.ERROR)
+    try:
+        export_ios_xcode_project(ctx = o3de_context,
+                                 tools_build_folder= args.tools_build_path,
+                                 ios_build_folder=args.ios_build_path,
+                                 should_build_tools = args.build_tools,
+                                 skip_asset_processing = args.skip_asset_processing,
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:117]:
```suggestion
        export_ios_xcode_project(ctx=o3de_context,
                                 tools_build_folder=args.tools_build_path,
                                 ios_build_folder=args.ios_build_path,
                                 should_build_tools=args.build_tools,
                                 skip_asset_processing=args.skip_asset_processing,
```

File: scripts/o3de/ExportScripts/export_source_ios_xcode.py [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+                                 logger = o3de_logger)
```
Comment (spham-amzn) [inline scripts/o3de/ExportScripts/export_source_ios_xcode.py:118]:
```suggestion
                                 logger=o3de_logger)
```

**Review by spham-amzn — APPROVED** [review]

**Review by cgalvan — APPROVED** [review]
